crates/sui-core/src/authority.rs

```diff
@@ -1301,22 +1295,78 @@ impl AuthorityState {
             .inc();
     }
 
-    /// Waits for fastpath (owned, package) dependency objects to become available,
-    /// until the specified max_wait timeout.
-    /// Returns true if all fastpath dependency objects are available, false otherwise.
+    /// Waits for fastpath (owned, package) dependency objects to become available.
+    /// Returns true if all fastpath dependency objects are available after an internal timeout, false otherwise.
+    /// Returns an error if a versioned input object will never become available.
     pub(crate) async fn wait_for_fastpath_dependency_objects(
         &self,
         transaction: &VerifiedTransaction,
         epoch: EpochId,
-        max_wait: Duration,
     ) -> SuiResult<bool> {
         let txn_data = transaction.data().transaction_data();
-        let fastpath_dependency_objects = txn_data.fastpath_dependency_objects()?;
-        if fastpath_dependency_objects.is_empty() {
-            return Ok(true);
+        let (move_objects, packages, receiving_objects) = txn_data.fastpath_dependency_objects()?;
+
+        // Gather and filter input objects to wait for.
+        // Skip waiting for object versions that already exist, and
+        // return an error if a newer version of an object is available.
+        let mut fastpath_dependency_objects = vec![];
+        for obj_ref in move_objects {
+            let (obj_id, cur_version, _digest) = obj_ref;
+            let Some(latest_obj_ref) = self
+                .get_object_cache_reader()
+                .get_latest_object_ref_or_tombstone(obj_id)
```

Comment:
Not very familiar with the underlying mechanics/semantics, but do we care if the returned `latest_obj_ref` designate a deleted object `(ObjectDigest == ObjectDigest::OBJECT_DIGEST_DELETED)`?